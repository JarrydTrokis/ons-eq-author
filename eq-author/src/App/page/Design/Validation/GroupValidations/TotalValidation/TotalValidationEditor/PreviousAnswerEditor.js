import React from "react";
import PropTypes from "prop-types";
import { get } from "lodash";

import { Field } from "components/Forms";

import { UnwrappedPreviousAnswerContentPicker as PreviousAnswerContentPicker } from "../../../PreviousAnswerContentPicker";

const PreviousAnswerEditor = ({ total, onChangeUpdate }) => (
  <Field>
    <PreviousAnswerContentPicker
      onSubmit={onChangeUpdate}
      selectedContentDisplayName={get(total.previousAnswer, "displayName")}
      data={total}
      path="availablePreviousAnswers"
      data-test="content-picker-select"
    />
  </Field>
);
PreviousAnswerEditor.propTypes = {
  total: PropTypes.shape({
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
  }).isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
};

export default PreviousAnswerEditor;
