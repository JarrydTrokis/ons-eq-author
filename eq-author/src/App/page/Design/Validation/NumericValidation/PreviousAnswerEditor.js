import React from "react";
import PropTypes from "prop-types";
import { get } from "lodash";
import ValidationError from "components/ValidationError";

import FieldWithInclude from "../FieldWithInclude";
import * as entityTypes from "constants/validation-entity-types";
import PreviousAnswerContentPicker from "../PreviousAnswerContentPicker";

import {
  ERR_REFERENCE_DELETED,
  ERR_REFERENCE_MOVED,
} from "constants/validationMessages";

export const errorMessages = {
  ERR_NO_VALUE: "Answer required",
  ERR_REFERENCE_DELETED,
  ERR_REFERENCE_MOVED,
};

const PreviousAnswerEditor = ({
  onChangeUpdate,
  answer,
  validation,
  readKey,
}) => {
  const { errors } = validation.validationErrorInfo;

  const errorCode = errors
    .map(e => e.errorCode)
    .find(errorCode => Object.keys(errorMessages).includes(errorCode));

  return (
    <FieldWithInclude
      id="inclusive"
      name="inclusive"
      onChange={onChangeUpdate}
      checked={validation.inclusive}
    >
      <PreviousAnswerContentPicker
        answerId={answer.id}
        onSubmit={onChangeUpdate}
        selectedContentDisplayName={get(
          validation.previousAnswer,
          "displayName"
        )}
        selectedId={get(validation.previousAnswer, "id")}
        path={`answer.validation.${readKey}.availablePreviousAnswers`}
        data-test="content-picker-select"
        hasError={errorCode}
      />
      {errorCode && (
        <ValidationError right={false}>
          {errorMessages[errorCode]}
        </ValidationError>
      )}
    </FieldWithInclude>
  );
};

PreviousAnswerEditor.propTypes = {
  validation: PropTypes.shape({
    id: PropTypes.string.isRequired,
    enabled: PropTypes.bool.isRequired,
    custom: PropTypes.number,
    inclusive: PropTypes.bool.isRequired,
    previousAnswer: PropTypes.shape({
      displayName: PropTypes.string.isRequired,
    }),
    entityType: PropTypes.oneOf(Object.values(entityTypes)),
  }).isRequired,
  answer: PropTypes.shape({
    id: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    properties: PropTypes.shape({
      unit: PropTypes.string,
    }),
    validationErrorInfo: PropTypes.shape({
      errors: PropTypes.arrayOf(
        PropTypes.shape({
          errorCode: PropTypes.string,
          field: PropTypes.string,
          id: PropTypes.string,
          type: PropTypes.string,
        })
      ),
    }),
  }).isRequired,
  readKey: PropTypes.string.isRequired,
  onChangeUpdate: PropTypes.func.isRequired,
};

export default PreviousAnswerEditor;
