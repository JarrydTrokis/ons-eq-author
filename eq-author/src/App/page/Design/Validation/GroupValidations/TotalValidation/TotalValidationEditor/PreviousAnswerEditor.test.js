import React from "react";
import { shallow } from "enzyme";

import PreviousAnswerEditor from "./PreviousAnswerEditor";

describe("Custom Editor", () => {
  let props;
  beforeEach(() => {
    props = {
      total: {
        id: "1",
        previousAnswer: {
          displayName: "My answer",
        },
      },
      onChangeUpdate: jest.fn(),
    };
  });

  it("should render", () => {
    expect(shallow(<PreviousAnswerEditor {...props} />)).toMatchSnapshot();
  });

  it("should trigger onChange when the number input changes", () => {
    const wrapper = shallow(<PreviousAnswerEditor {...props} />);
    wrapper
      .find("[data-test='content-picker-select']")
      .simulate("submit", { name: "previousAnswer", value: { id: "1" } });
    expect(props.onChangeUpdate).toHaveBeenCalledWith({
      name: "previousAnswer",
      value: { id: "1" },
    });
  });
});
