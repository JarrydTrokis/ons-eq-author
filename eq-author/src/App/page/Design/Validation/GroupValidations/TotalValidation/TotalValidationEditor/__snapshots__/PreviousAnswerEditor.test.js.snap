// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Editor should render 1`] = `
<Field
  disabled={false}
  last={false}
>
  <UnwrappedPreviousAnswerContentPicker
    data={
      Object {
        "id": "1",
        "previousAnswer": Object {
          "displayName": "My answer",
        },
      }
    }
    data-test="content-picker-select"
    onSubmit={[MockFunction]}
    path="availablePreviousAnswers"
    selectedContentDisplayName="My answer"
  />
</Field>
`;
